feat: support `bn256_pairing` precompiled contract by Godwoken syscall

Flouse opened the pull request:
Previously, `bn256_pairing` precompiled contract was not supported due to too high cycle cost.

This PR implements `bn256_pairing` precompiled contract by introducing new Godwoken syscall:
```markdown
sys_bn_pairing
```

## Dependencies
- https://github.com/nervosnetwork/godwoken/pull/808
- https://github.com/nervosnetwork/godwoken-scripts/pull/134


## Tests

test code:
https://github.com/Flouse/godwoken-polyjuice/blob/47865fc349d9485021b3e75205cfd621c1ada8e1/polyjuice-tests/src/test_cases/pre_compiled_contracts.rs#L163-L697

cmd:
```bash
cd polyjuice-tests/
cargo test --release -- test_bn256_add --nocapture 2>&1
cargo test --release -- test_bn256_scalar_mul --nocapture 2>&1
cargo test --release -- test_bn256_pairing --nocapture 2>&1
```

| Test Case                      | Elapsed Time (ms) | Used Gas | Used Cycles (virtual+execution) |
| ------------------------------ | ----------------: | -------: | ------------------------------: |
| Bn256Add                       |                 1 |     ~27K |                           ~693K |
| Bn256ScalarMul                 |                <2 |     ~31K |                           ~690K |
| Bn256Pairing one_point         |                 5 |    ~105K |                           ~918K |
| Bn256Pairing two_point_match_2 |                 6 |    ~141K |                          ~1024K |
| Bn256Pairing two_point_match_3 |                 7 |    ~141K |                          ~1024K |
| Bn256Pairing two_point_match_4 |                 8 |    ~141K |                          ~1024K |
| Bn256Pairing ten_point_match_1 |                22 |    ~432K |                          ~1874K |
| Bn256Pairing ten_point_match_2 |                23 |    ~436K |                          ~1874K |
| Bn256Pairing ten_point_match_3 |                 8 |    ~142K |                          ~1024K |

Flouse committed:
feat: Add bn256_pairing precompiled contract

Add bn submodule - Pairing cryptography library

Flouse committed:
Fix stack pointer not aligned in ckb-vm

- https://github.com/nervosnetwork/ckb-vm/issues/97
- https://github.com/nervosnetwork/ckb-c-stdlib/pull/33

Flouse committed:
fix(fuzz): build libalt_bn128.o in Makefile of fuzz test

fix(fuzz): unknown register name 't0' in asm

https://github.com/Flouse/godwoken-polyjuice/runs/3676372772?check_suite_focus=true#step:10:14

Flouse committed:
fix(CI): install riscv64imac-unknown-none-elf in build/libalt_bn128.a

error[E0463]: can't find crate for `core`
  |
  = note: the `riscv64imac-unknown-none-elf` target may not be installed
  = help: consider downloading the target with `rustup target add riscv64imac-unknown-none-elf`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

TheWaWaR commented on polyjuice-tests/src/helper.rs:
```diff
@@ -359,7 +359,12 @@ impl PolyjuiceArgsBuilder {
+    // If you want to watch the [contract debug] logs in Polyjuice,
+    // please change the log level from `info` to `debug`.
+    // then run `cargo test -- [test_filter] --nocapture`, 
+    // or run `RUST_LOG=debug cargo test -- [test_filter] --nocapture` directly
```
better with `RUST_LOG=gw=debug`, the `gw` prefix can filter out a lot unrelated log.

Flouse committed:
Update helper.rs

the gw prefix can filter out a lot unrelated log

XuJiandong commented on c/contracts.h:
```diff
@@ -5,12 +5,19 @@
+ * alt_bn128_staticlib
+ * @see https://github.com/mohanson/bn/blob/d1c73cc198a14ca894a37ba99df88a9fb812a43c/alt_bn128_staticlib/src/ethereum.rs
+ */
+extern "C" uint32_t alt_bn128_add(const uint8_t *data, uint32_t data_len, uint8_t *output);
```
It's better to return "uint64_t" or "int" as the error code.

Flouse replied:
The extern "C" functions use `u32` as a return code:
https://github.com/mohanson/bn/blob/master/alt_bn128_staticlib/src/ethereum.rs#L2

should I use int in c/contracts.h?

XuJiandong replied:
it seems `u32` is OK here.

XuJiandong commented on c/generator.c:
```diff
@@ -13,4 +13,12 @@
-int main() { return run_polyjuice(); }
+int main() {
+  // A temporal patch to solve https://github.com/nervosnetwork/ckb-vm/issues/97
+  CKB_SP_ALIGN;
```
This align can be removed, I think.

Flouse committed:
Remove the temporal patch of CKB_SP_ALIGN

## References:
- https://github.com/nervosnetwork/ckb-c-stdlib/pull/33
- https://github.com/nervosnetwork/ckb-vm/issues/97

Flouse committed:
deps: remove submodule bn128

Flouse committed:
git rm --cached deps/ckb-binary-patcher

Flouse committed:
docs: update README.md

Flouse committed:
feat: support bn256_pairing precompiled contract by Godwoken syscall

Flouse committed:
deps: rm pairing cryptography library bn

Flouse committed:
test: add test cases for EIP-196 and EIP-197

Flouse committed:
chore(CI): Run Ethereum Test in a single step

Flouse committed:
chore: tune optimization level

Flouse committed:
test: refactor `check_cycles` in helper.rs

Flouse committed:
chore: tune compiler flags

Flouse committed:
test: fix multicall3

Flouse committed:
fix(test): test_cases/pre_compiled_contracts.rs

Flouse committed:
Bump POLYJUICE_VERSION  to 1.5.0

Flouse committed:
perf: activate compiler optimization for size

jjyr commented:
Please rebase
